petri/src/tracing.rs

```diff
@@ -1,100 +1,302 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-use std::fs::File;
+use fs_err::File;
+use fs_err::PathExt;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::io::Write as _;
 use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
 use tracing::level_filters::LevelFilter;
+use tracing::Level;
 use tracing_subscriber::filter::Targets;
 use tracing_subscriber::fmt::format::FmtSpan;
-use tracing_subscriber::fmt::writer::EitherWriter;
-use tracing_subscriber::fmt::writer::Tee;
 use tracing_subscriber::fmt::MakeWriter;
-use tracing_subscriber::fmt::TestWriter;
 use tracing_subscriber::layer::SubscriberExt;
 use tracing_subscriber::util::SubscriberInitExt;
 
-pub(crate) const LINUX_TARGET: &str = "linux_log";
-pub(crate) const PCAT_TARGET: &str = "pcat_log";
-pub(crate) const UEFI_TARGET: &str = "uefi_log";
-pub(crate) const OPENHCL_TARGET: &str = "openhcl_log";
-pub(crate) const OPENVMM_TARGET: &str = "openvmm_log";
+/// A source of [`PetriLogFile`] log files for test output.
+#[derive(Clone)]
+pub struct PetriLogSource(Arc<LogSourceInner>);
 
-pub(crate) fn try_init_tracing(
-    log_file: File,
-) -> Result<(), tracing_subscriber::util::TryInitError> {
+struct LogSourceInner {
+    root_path: PathBuf,
+    json_log: JsonLog,
+    log_files: Mutex<HashMap<String, PetriLogFile>>,
+    attachments: Mutex<HashMap<String, u64>>,
+}
+
+impl PetriLogSource {
+    /// Returns a log file for the given name.
+    ///
+    /// The name should not have an extension; `.log` will be appended
+    /// automatically.
+    pub fn log_file(&self, name: &str) -> anyhow::Result<PetriLogFile> {
+        use std::collections::hash_map::Entry;
+
+        let mut log_files = self.0.log_files.lock();
+        let log_file = match log_files.entry(name.to_owned()) {
+            Entry::Occupied(occupied_entry) => occupied_entry.get().clone(),
+            Entry::Vacant(vacant_entry) => {
+                let mut path = self.0.root_path.join(name);
+                path.set_extension(".log");
```

Comment:
might want to preserve the context around why ".log" is a better choice than ".txt" (namely, the WSL2 / Windows-Defender interplay).